lib/internal/repl.js

```diff
@@ -162,62 +160,25 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
         }
       }
     }
-
-    fs.open(historyPath, 'w', onhandle);
-  }
-
-  function onhandle(err, hnd) {
-    if (err) {
-      return ready(err);
-    }
-    repl._historyHandle = hnd;
-    repl.on('line', online);
-
-    // reading the file data out erases it
-    repl.once('flushHistory', function() {
-      repl.resume();
-      ready(null, repl);
-    });
-    flushHistory();
-  }
-
-  // ------ history listeners ------
-  function online() {
-    repl._flushing = true;
-
-    if (timer) {
-      clearTimeout(timer);
-    }
-
-    timer = setTimeout(flushHistory, kDebounceHistoryMS);
+    // Flush history on close
+    repl.on('close', flushHistory);
+    repl.resume();
+    ready(null, repl);
   }
 
   function flushHistory() {
-    timer = null;
-    if (writing) {
-      pending = true;
-      return;
-    }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
-      }
+    const data = fs.readFileSync(historyPath, 'utf8');
+    const newLines = repl.history.length - initialHistoryLength;
+    let history = repl.history.slice(0, newLines);
+    if (data) {
+      history = history.concat(parseData(data)).slice(0, repl.historySize);
     }
+    const fd = fs.openSync(historyPath, 'a');
+    fs.writeSync(fd, history.join('\n'), 0, 'utf8');
```

Comment:
@addaleax that's not actually the case. Since the history file is read in its entirety and rewritten each time, a trailing newline is not necessary. The pattern is:

  * On startup read the history file into memory and split on newline to create the in-memory history array.
  * Keep track of the number of lines found in the original history file.
  * On shutdown, extract the new lines added to the in-memory history.
  * Re-read the history file and concatenate the file data with the new lines from our in-memory history.
  * Write the entire history file with the data that was read from the file, plus the new lines that have been concatenated to it.
  * Close the history file.

No newline required! Also, note that the final `fs.openSync` used to write the history file is opened with the `a` flag, to get around the Windows issues with hidden files. When I do the subsequent `fs.write`, the write just begins at position `0` in the file.